### ISSUE TITLE
Split passthrough args to specific pipelines

### ISSUE THREAD

**Issue by anthonyshew** [vercel/turbo#5743]
### Which project is this feature idea for?

Turborepo

### Describe the feature you'd like to request

Currently, [the documentation](https://turbo.build/repo/docs/reference/command-line-reference/run) shows that passthrough args are possible when calling many pipelines. However, this introduces a partitioning problem for the cache since **all** arguments are applied to **all** pipelines.

As an example, let's say I use `turbo lint format` where `lint` is an ESLint script and `format` is a Prettier script.

If I want to run a "clean up all my code" command, I would probably write `turbo lint format -- --write --fix`.
- `--fix` is responsible for telling `turbo lint` (ESLint under the hood) to fix any issues that it can and write them to disk
- `--write` is responsible for telling `turbo format` (Prettier under the hood) to write formatting fixes to disk

This works at first glance but has a hidden downside: The formatting command effectively becomes `turbo format -- --write --fix`, unnecessarily including the `--fix` argument. The same is true for `turbo lint -- --write --fix` (erroneous `--write`).

If I were to then run `turbo lint -- --fix` (the correct command), I will now have cache misses because the first command I called was partitioned using both arguments rather than the one I actually need.

This API interface blocks me from being able to do what I truly want: Run multiple pipelines with their own arguments in parallel with one `turbo` command.

---

Note that `lint`, `format`, ESLint, and Prettier are arbitrary. ESLint throws errors on unknown args so `eslint . --fix --write` would throw an error on the `--write` arg. I'm only trying to demonstrate the problem for the API surface.

### Describe the solution you'd like

I admittedly haven't thought of a truly elegant solution for this yet. I'm mostly drawing up this issue as a conversation starter. Only thing coming to mind for me thinking about it for a few minutes is something like:

`turbo [lint -- --fix] [format -- --write] -- --foo`

With this syntax, we would get:
- `--fix` and `--foo` as args for the `lint` pipeline
- `--write` and `--foo` as args for the `format` pipeline

### Describe alternatives you've considered

Currently, my workaround is to run the commands serially. `turbo lint -- --fix && turbo format -- --write` - which, of course, does not achieve the desired parallelism. But at least I can get everything done in one command!

I'll have a longer think about alternatives; I may be back! 😄

**Comment by nathanhammond** [comment]
I never recommend using command line arguments like this and instead recommend configuring a separate task inside of `turbo.json`. As you noted there are many downsides to using passthrough flags beyond just "is not easily shareable between team members."

The graph can easily be hundreds of items and there really isn't a scalable pattern to pass CLI flags from the `turbo` invocation itself. I would generally be more-willing to remove the ability to pass arguments in like this at all than add the ability to address which task a particular flag should be passed to.

We have a way to control flags on a per-task basis already, and control how they execute inside the graph, via `turbo.json`:
- Scales fairly well.
- Shareable configuration.
- No unnecessary cache partitioning.
- Tiny bit more configuration for including the additional tasks in the graph. (once, not every single invocation)
- No new CLI microsyntax.

For small projects the ease of `-- --fix --write` may be better than `turbo run build eslint-fix prettier-write` but precomposing the will give better and more-consistent results long term.

**Comment by anthonyshew** [comment]
I had considered writing these args directly into the scripts but was looking for a way to deduplicate the configuration. With the approach you're describing, all of my `package.json`'s will need to have:

```json
{
  "scripts": {
    "format": "prettier . --check --cache --cache-location='node_modules/.cache/prettiercache'",
    "format:fix": "prettier . --write --cache --cache-location='node_modules/.cache/prettiercache'",
    "lint": "eslint . --cache --cache-location 'node_modules/.cache/.eslintcache'",
    "lint:fix": "eslint . --fix --cache --cache-location 'node_modules/.cache/.eslintcache'"
  }  
}
```

I was hoping to be able to be able to write those once instead of into each workspaces' scripts. 🤔 I see the tradeoffs, though, so might be wishful thinking.

---

> We have a way to control flags on a per-task basis already, and control how they execute inside the graph, via turbo.json

Here, are you describing something like:

```json
{
  "$schema": "https://turbo.build/schema.json",
  "pipeline": {
    "format": {
      "dependsOn": ["^topo"],
      "outputs": ["node_modules/.cache/.prettiercache"]
      "args": "--cache --cache-location 'node_modules/.cache/.prettiercache'"
      },
    "format:fix": {
      "dependsOn": ["^topo"],
      "outputs": ["node_modules/.cache/.prettiercache"]
      "args": "--write --cache --cache-location 'node_modules/.cache/.prettiercache'"
      }
  }
}
```

That seems like it would be rather ergonomic at first glance to me. Wondering what I'm not considering?

**Comment by anthonyshew** [comment]
After working with your strategy for awhile, it's clearly better. Thanks for the tip.

I still wonder about getting to place default args into `turbo.json` but I can see plenty of potential problems with that approach as well. Having the scripts be their own source of truth does seem to be the most straightforward and scalable approach.

**Comment by nebez** [comment]
@anthonyshew @nathanhammond - can you point me to the docs for this `args` param you're alluding to in your examples? I can't find it in https://turbo.build/schema.json or the [docs](https://turbo.build/repo/docs/reference/command-line-reference/run).

This would help me greatly. I'm currently doing some sharding of vitest runners thru turbo, and getting cache misses on all upstream build steps is annoying.

**Comment by anthonyshew** [comment]
The `args` I was proposing was only an idea and not a part of the API today.
